enhancements/api-review/add-new-CRD-ImageSourceDigestPolicy-and-ImageSourceTagPolicy-to-config.openshift.io.md

```diff
@@ -0,0 +1,403 @@
+
+### Implementation Details/Notes/Constraints [optional]
+
+Implementing this enhancement requires changes in:
+- [containers/images](https://github.com/containers/image): registries.conf needs change its mirror setting to make digests only configurable for mirrors registry, not for the main registry. This can be achieved by having a boolean for each mirror table, or having separated mirror lists: a list of mirrors that allow for tag specification,
+and a list of mirrors require digest specification, or have different boolean value.
+- [openshift/api/config/v1](https://github.com/openshift/api/tree/master/config/v1): definition the schema and API of the ImageSourceDigestPolicy and ImageSourceTagPolicy.
+- [openshift/client-go](https://github.com/openshift/client-go), [openshift/cluster-config-operator](https://github.com/openshift/cluster-config-operator/pull/220): rebase the openshift/api version in these repositories to apply the new CRD
+to the cluster.
+- [openshift/runtime-utils/pkg/registries](https://github.com/openshift/runtime-utils/tree/master/pkg/registries): helper functions to edit registries.conf.
+- [openshift/machine-config-operator](https://github.com/openshift/machine-config-operator): MCO needs watch for the ImageSourceDigestPolicy and ImageSourceDigestPolicy. The machine-config-operator/pkg/controller/container-runtime-config controller needs to operate the ImageSourceDigestPolicy and ImageSourceDigestPolicy CRDs.
+Converts the existing ImageContenSourcePolicy objects to objects of new CRD.
+- This [document](https://docs.google.com/document/d/11FJPpIYAQLj5EcYiJtbi_bNkAcJa2hCLV63WvoDsrcQ/edit?usp=sharing) keeps a list of components that use operator.openshift.io/v1alpha1 ImageContentSourcePolicy. Need to change those repositories to upgrade to ImageSourceDigestPolicy.
+
+#### Notes
+
+1. During the upgrade path, MCO can watch for both old CR ImageContentSourcePolicy and new CRs and create objects.
+
+2. The merge order of mirrors for the same source is deterministic by topological sorting:<br/>
+Order is preserving the relative order of the mirrors using topological sorting. A graph is formed using each mirror as a
+vertex, adding an unidirectional edge between adjacent mirrors from the same mirror list. If the graph is cyclic graph, the topological sort will start
+from the first vertex after the alphabetical sorting of the vertces. If multiple verteces is from the same vertex, they will be
+visited in alphabetical order.
+
+3. The ImageStream does not supported pulling using tags yet, however, work has started on this feature.
+
+### Risks and Mitigations
+
+For users that opt into using this feature, pulling images from mirror registries without the digest specifications could lead to returning different image version from different registry if the image tag mapping is out of sync. However, OpenShift core is not exposed to this risk because it requires
```

Comment:
```suggestion
For users that opt into using this feature, pulling images from mirror registries without the digest specifications could lead to returning a different image version from a different registry if the image tag mapping is out of sync. However, OpenShift core is not exposed to this risk because it requires
```